file may contain typos

# Installing & using Julia

## Installing

https://julialang.org/downloads/

chose your platform and follow [[help]](https://julialang.org/downloads/platform/#linux_and_freebsd)


Linux: 

>wget https://julialang-s3.julialang.org/bin/linux/x64/1.8/julia-1.8.2-linux-x86_64.tar.gz

>tar zxvf julia-1.8.2-linux-x86_64.tar.gz

## Workflow options :

1. Terminal
    1. scripts in files.jl
    2. REPL
2. Jupyter lab or notebook (needs additional installation of `IJulia` package)
    1. Cells
3. Visual Studio Code
    1. Text editor + terminal
4. others

### small Julia tips

Julia supports latex-like symbols 

instead of `alpha` one can use \alpha\<TAB\> to get $\alpha$

In [12]:
alpha = 1.0
α = 2.0
α > alpha 

true

Even more!  nice indices

$\sigma^x$ = \sigma\<TAB\>\\^x\<TAB\>

In [13]:
σˣ = [0 1; 1 0]

2×2 Matrix{Int64}:
 0  1
 1  0

### `help` and `Pkg`
Julia has good `help` and package manager `Pkg`.
For help hit `?` button and type a keyword 

In [14]:
? random

search: lowrankdowndate lowrankdowndate! RankDeficientException

Couldn't find random
Perhaps you meant rand, randn, rank, range, nand, tand or rad2deg


No documentation found.

Binding `random` does not exist.


In [15]:
? rand

search: rand randn RankDeficientException lowrankdowndate lowrankdowndate!



```
rand([rng=GLOBAL_RNG], [S], [dims...])
```

Pick a random element or array of random elements from the set of values specified by `S`; `S` can be

  * an indexable collection (for example `1:9` or `('x', "y", :z)`),
  * an `AbstractDict` or `AbstractSet` object,
  * a string (considered as a collection of characters), or
  * a type: the set of values to pick from is then equivalent to `typemin(S):typemax(S)` for integers (this is not applicable to [`BigInt`](@ref)), to $[0, 1)$ for floating point numbers and to $[0, 1)+i[0, 1)$ for complex floating point numbers;

`S` defaults to [`Float64`](@ref). When only one argument is passed besides the optional `rng` and is a `Tuple`, it is interpreted as a collection of values (`S`) and not as `dims`.

!!! compat "Julia 1.1"
    Support for `S` as a tuple requires at least Julia 1.1.


# Examples

```julia-repl
julia> rand(Int, 2)
2-element Array{Int64,1}:
 1339893410598768192
 1575814717733606317

julia> using Random

julia> rand(MersenneTwister(0), Dict(1=>2, 3=>4))
1=>2

julia> rand((2, 3))
3

julia> rand(Float64, (2, 3))
2×3 Array{Float64,2}:
 0.999717  0.0143835  0.540787
 0.696556  0.783855   0.938235
```

!!! note
    The complexity of `rand(rng, s::Union{AbstractDict,AbstractSet})` is linear in the length of `s`, unless an optimized method with constant complexity is available, which is the case for `Dict`, `Set` and dense `BitSet`s. For more than a few calls, use `rand(rng, collect(s))` instead, or either `rand(rng, Dict(s))` or `rand(rng, Set(s))` as appropriate.



# Packages and installation

To install a package, e.g. `Plots`, in julia REPL (Read-Eval-Print loop) terminal hit `]` button and type
`add Plots`

or equivalently
`import Pkg; Pkg.add("Plots")`.

To use it 
`using PkgName`

In [16]:
using Random
rnd = rand()

0.42565473370306184

In [17]:
rnd_v = rand(5)

5-element Vector{Float64}:
 0.8174372582614207
 0.8267084202906579
 0.18161429809076057
 0.6301576312669225
 0.05822246395085506

In [18]:
rnd_m = rand(5,5)

5×5 Matrix{Float64}:
 0.302954  0.839343  0.0753287  0.00448834  0.856237
 0.952945  0.221199  0.492158   0.206204    0.309418
 0.166847  0.649823  0.253243   0.777478    0.368335
 0.361857  0.739037  0.250039   0.0603737   0.772568
 0.558281  0.706662  0.534584   0.192844    0.845517

# Quantum Many-body physics

### One qbit state
$|\psi\rangle = |1\rangle = |\uparrow\rangle = 
    \begin{bmatrix}
        1 \\
        0 \\
    \end{bmatrix}$
    
$|\psi\rangle = |0\rangle = |\downarrow\rangle = 
    \begin{bmatrix}
        0 \\
        1 \\
    \end{bmatrix}$ 
 

### Many qbit system
$|\psi\rangle = |1 0\rangle = |\uparrow \downarrow\rangle = 
    \begin{bmatrix}
        1 \\
        0 \\
    \end{bmatrix} \otimes
    \begin{bmatrix}
        0 \\
        1 \\
    \end{bmatrix}
    = \begin{bmatrix}
        1 * 
        \begin{bmatrix}
            0 \\
            1 \\
        \end{bmatrix}
        \\ 0 * 
        \begin{bmatrix}
            0 \\
            1 \\
        \end{bmatrix}
    \end{bmatrix} 
    =
    \begin{bmatrix}
            0 \\
            1 \\
            0 \\
            0 \\
        \end{bmatrix}
    $   

$\otimes$ = \otimes\<TAB\>

In [19]:
⊗(x, y) = kron(x, y)

⊗ (generic function with 1 method)

In [20]:
up = [1, 0]; # ; allows to suppress output
down = [0, 1];
display(up ⊗ down)
display(down ⊗ up)
println("|10⟩ = ", up ⊗ down)
println("|01⟩ = ", down ⊗ up)

4-element Vector{Int64}:
 0
 1
 0
 0

4-element Vector{Int64}:
 0
 0
 1
 0

|10⟩ = [0, 1, 0, 0]
|01⟩ = [0, 0, 1, 0]


### Entangled state

$|\psi\rangle = \frac{|1 0\rangle + |0 1\rangle}{\sqrt{2}} = \frac{|\uparrow \downarrow\rangle + |\downarrow\uparrow\rangle}{\sqrt{2}} $ 


$\begin{bmatrix}
           0 \\
           1/\sqrt{2} \\
           0 \\
           0 \\
         \end{bmatrix}$ +
$\begin{bmatrix}
           0 \\
           0 \\
           1/\sqrt{2} \\
           0 \\
         \end{bmatrix}$ =
$\begin{bmatrix}
           0 \\
           1/\sqrt{2} \\
           1/\sqrt{2} \\
           0 \\
         \end{bmatrix}$

In [21]:
using LinearAlgebra

ent_state = up ⊗ down + down ⊗ up
ent_state /= norm(ent_state)
ent_state

4-element Vector{Float64}:
 0.0
 0.7071067811865475
 0.7071067811865475
 0.0

### Many body product states

In [22]:
function nbodyWF(n, v)
    state = fill(v, n) # array of 2 dim vectors
    result = foldl(⊗, state) # product state, e.g. |..↓↓↓↓..⟩ state
    return result
end

nbodyWF (generic function with 1 method)

In [23]:
nbodyWF(4, down);

# Quantum many body operators

$n |1\rangle = 1 |1\rangle$

$n |0\rangle = 0 |1\rangle$

$\begin{bmatrix}
           1 & 0 \\
           0 & 0 \\
         \end{bmatrix}
\begin{bmatrix}
           1 \\
           0 \\
         \end{bmatrix}=
         1 \times
\begin{bmatrix}
           1 \\
           0 \\
         \end{bmatrix}$

$\begin{bmatrix}
           1 & 0 \\
           0 & 0 \\
         \end{bmatrix}
\begin{bmatrix}
           0 \\
           1 \\
         \end{bmatrix}=
\begin{bmatrix}
           0 \\
           0 \\
         \end{bmatrix} = 0 \times
\begin{bmatrix}
           0 \\
           1 \\
         \end{bmatrix}$
         
$\langle 1| n |1\rangle = 1 \langle 1|1\rangle = 1$

$\langle 0| n |0\rangle = 0 \langle 0|0\rangle = 0$

In [24]:
n = [1 0; 0 0];
println("⟨1|n|1⟩ = ", up' * n * up) 
println("⟨0|n|0⟩ = ",down' * n * down)

⟨1|n|1⟩ = 1
⟨0|n|0⟩ = 0


### 2 qbit system

$n_1 = n \otimes id$  -  acts only on the 1st qbit

$n_2 = id \otimes n$  -  acts only on the 2nd qbit

In [25]:
id = [1 0; 0 1];
n1 = n ⊗ id;
n2 = id ⊗ n;
updown = up ⊗ down;
upup = up ⊗ up;
downdown = down ⊗ down;
downup = down ⊗ up;

In [26]:
println("⟨01|n₁|01⟩ = ", downup' * n1 * downup)
println("⟨10|n₁|10⟩ = ", updown' * n1 * updown)
println("⟨10|n₂|10⟩ = ", updown' * n2 * updown)
println("⟨11|n₂|11⟩ = ", upup' * n2 * upup)

⟨01|n₁|01⟩ = 0
⟨10|n₁|10⟩ = 1
⟨10|n₂|10⟩ = 0
⟨11|n₂|11⟩ = 1


In [27]:
println("⟨01|n₁+n₂|01⟩ = ", downup' * (n1+n2) * downup)
println("⟨10|n₁+n₂|10⟩ = ", updown' * (n1+n2) * updown)
println("⟨11|n₁+n₂|11⟩ = ", upup' * (n1+n2) * upup)
println("⟨00|n₁+n₂|00⟩ = ", downdown' * (n1+n2) * downdown)

⟨01|n₁+n₂|01⟩ = 1
⟨10|n₁+n₂|10⟩ = 1
⟨11|n₁+n₂|11⟩ = 2
⟨00|n₁+n₂|00⟩ = 0


### How to construct a many body operator ?
$n_i = id\otimes ... id\otimes n \otimes id ...\otimes id$

Total number of excited qbits
$N = \sum n_i$

#### Intecations in quantum systems
$\sigma^x |0\rangle = |1\rangle$

$\sigma^x |1\rangle = |0\rangle$

$\begin{bmatrix}
           0 & 1 \\
           1 & 0 \\
         \end{bmatrix}
\begin{bmatrix}
           1 \\
           0 \\
         \end{bmatrix}=
\begin{bmatrix}
           0 \\
           1 \\
         \end{bmatrix}$


$\begin{bmatrix}
           0 & 1 \\
           1 & 0 \\
         \end{bmatrix}
\begin{bmatrix}
           0 \\
           1 \\
         \end{bmatrix}=
\begin{bmatrix}
           1 \\
           0 \\
         \end{bmatrix}$


$\sigma^x_i = ... id\otimes \sigma^x \otimes id ...$

$n_i n_j = ... id\otimes n \otimes id ...id\otimes n \otimes id ...$

### Hamiltonian and time dynamics of quantum systems

Hamiltonian == Energy function == cost function

$H = -\sum n_i$

or

$H = \sum n_i n_j$

or 

$H(t) = \alpha(t)\sum n_i n_j - \beta(t)\sum n_i$ 

or

$H(t) = \sum \alpha_{ij} n_i n_j - \delta(t)\sum n_i + \frac{\Omega(t)}{2}\sum \sigma^x_i$ .



In [28]:
⊗(x, y) = kron(x, y)

delta(x::Int, y::Int)= x == y ? 1.0 : 0.0

using SparseArrays

id = [1 0; 0 1]
σˣ = [0 1; 1 0]
ProjUp = [1 0; 0 0]

function Op1(L, i, σ)
    operator_string = fill(id, L)
    operator_string[i] = σ 
    return foldl(⊗, operator_string)
end

function Op2(L, i, σ1, j, σ2)
    operator_string = fill(id, L)
    operator_string[i] = σ1 
    operator_string[j] = σ2 
    return foldl(⊗, operator_string)
end

function H(L, Ω, δ)
    H = zeros(2^L, 2^L) ## = fill(0, 2^N, 2^N)
    C_6 = 5420158.53
    for i in 1:L
        for j in i+1:L
            if j == i + 1
                H += Op2(L, i, ProjUp, j, ProjUp)#iteratively applies ⊗ to every element of vector 'interaction' 
            end
        end
        H += 0.5 * Ω * Op1(L,i, σˣ)
        H -= 1.0 * δ * Op1(L,i, ProjUp)
    end
    return H #sparse(H)
end

H(3, 0.0, 25.)

8×8 Matrix{Float64}:
 -73.0    0.0    0.0    0.0    0.0    0.0    0.0  0.0
   0.0  -49.0    0.0    0.0    0.0    0.0    0.0  0.0
   0.0    0.0  -50.0    0.0    0.0    0.0    0.0  0.0
   0.0    0.0    0.0  -25.0    0.0    0.0    0.0  0.0
   0.0    0.0    0.0    0.0  -49.0    0.0    0.0  0.0
   0.0    0.0    0.0    0.0    0.0  -25.0    0.0  0.0
   0.0    0.0    0.0    0.0    0.0    0.0  -25.0  0.0
   0.0    0.0    0.0    0.0    0.0    0.0    0.0  0.0

Ground state(GS) is a state with minimal energy 

$H |GS\rangle = E_{min} |GS\rangle$

or

$\langle GS |H |GS\rangle = E_{min} $
## We can make a brute force search and find a ground state !

In [29]:
function ψ(i,L)
    bit_string = string(i-1; base = 2, pad = L)
    state = fill([0, 1], L) # |..↓↓↓↓..⟩

    for i in 1:L
        if (bit_string[i] == '1')
            state[i] = [1, 0]
        end
    end
    psi = foldl(⊗, state)
    return psi, bit_string
end

ψ (generic function with 1 method)

In [30]:
function min_ψ(Ham)
    result = Vector{Tuple{Any,Any}}()
    L = Int(log2(size(Ham,1)))
    for i in 1: 2^L
        psi, bit_string = ψ(i, L)    
        en = psi' * Ham * psi
        push!(result, (en,bit_string))
    end
    return sort(result)
end

min_ψ (generic function with 1 method)

In [31]:
sort(min_ψ(H(5, 0.0, 1.0)))

32-element Vector{Tuple{Any, Any}}:
 (-3.0, "10101")
 (-2.0, "00101")
 (-2.0, "01001")
 (-2.0, "01010")
 (-2.0, "01011")
 (-2.0, "01101")
 (-2.0, "10001")
 (-2.0, "10010")
 (-2.0, "10011")
 (-2.0, "10100")
 (-2.0, "10110")
 (-2.0, "10111")
 (-2.0, "11001")
 ⋮
 (-1.0, "00110")
 (-1.0, "00111")
 (-1.0, "01000")
 (-1.0, "01100")
 (-1.0, "01110")
 (-1.0, "01111")
 (-1.0, "10000")
 (-1.0, "11000")
 (-1.0, "11100")
 (-1.0, "11110")
 (-1.0, "11111")
 (0.0, "00000")

# Time evolution

Solution of the Schrodinger equation

$|\psi(t)\rangle = e^{-iHt}|\psi(0)\rangle$

![Pulse](./tutorials_afm_prep_8_0.png)

In [32]:
#import Pkg; Pkg.add("JSON")
using LinearAlgebra
include("./JSONparser.jl")

dataJSON = parse_json("./pulse9.json")
coordinates = dataJSON.coordinates
L = length(coordinates)


function H(coordinates, Ω, δ)
    L = length(coordinates)
    H = zeros(2^L, 2^L) ## = fill(0, 2^N, 2^N)
    C_6 = 5420158.53
    for i in 1:L
        xi = coordinates[i].x
        yi = coordinates[i].y
        for j in i+1:L
            xj = coordinates[j].x
            yj = coordinates[j].y
            
            rij = sqrt((xi - xj)^2 + (yi - yj)^2)
            coeff = C_6 / rij^6
            
            H += coeff * Op2(L, i, ProjUp, j, ProjUp)#iteratively applies ⊗ to every element of vector 'interaction' 
        end
        H += 0.5 * Ω * Op1(L,i, σˣ)
        H -= 1.0 * δ * Op1(L,i, ProjUp)
    end
    return H #sparse(H)
end


H (generic function with 1 method)

In [40]:
dt = 100. # [nanoseconds] #execution time ≈ 36 sec

pulsesequence = discretize(dataJSON.pulse, dt);

length(pulsesequence)

100

In [41]:
using KrylovKit: exponentiate

psi, bits = ψ(1, L)

tonanoseconds = 1e-3
dt *= tonanoseconds 
function run_pulse(pulsesequence, psi)
    for (i, pulse) in enumerate(pulsesequence)
        println("step $i/$(length(pulsesequence)) ")
        Ω = pulse.omega
        δ = pulse.delta
        Ham = H(coordinates, Ω, δ);
        #psi = exp(-1.0im * Ham * dt) * psi
        psi, info = exponentiate(Ham, -1.0im * dt, psi)
        
    end
    return psi
end

@time(psi = run_pulse(pulsesequence, psi)); # ≈ 36 sec



step 1/100 
step 2/100 
step 3/100 
step 4/100 
step 5/100 
step 6/100 
step 7/100 
step 8/100 
step 9/100 
step 10/100 
step 11/100 
step 12/100 
step 13/100 
step 14/100 
step 15/100 
step 16/100 
step 17/100 
step 18/100 
step 19/100 
step 20/100 
step 21/100 
step 22/100 
step 23/100 
step 24/100 
step 25/100 
step 26/100 
step 27/100 
step 28/100 
step 29/100 
step 30/100 
step 31/100 
step 32/100 
step 33/100 
step 34/100 
step 35/100 
step 36/100 
step 37/100 
step 38/100 
step 39/100 
step 40/100 
step 41/100 
step 42/100 
step 43/100 
step 44/100 
step 45/100 
step 46/100 
step 47/100 
step 48/100 
step 49/100 
step 50/100 
step 51/100 
step 52/100 
step 53/100 
step 54/100 
step 55/100 
step 56/100 
step 57/100 
step 58/100 
step 59/100 
step 60/100 
step 61/100 
step 62/100 
step 63/100 
step 64/100 
step 65/100 
step 66/100 
step 67/100 
step 68/100 
step 69/100 
step 70/100 
step 71/100 
step 72/100 
step 73/100 
step 74/100 
step 75/100 
step 76/100 
step 77/100 
step 78/

In [44]:
density = Vector{Any}()
for i in 1:L
    ni = psi' * Op1(L,i,ProjUp) * psi
    r_ni = round(real(ni); digits = 2)
    push!(density, r_ni)
end
reshape(density, 3, 3)

3×3 Matrix{Any}:
 0.99  0.0   0.99
 0.0   1.0   0.0
 1.0   0.01  1.0

In [49]:
bitstrings = Vector{Tuple{Any,Any}}()
for i in 1:2^L
    psi0, bitstring = ψ(i, L)
    p = round(abs2(psi0' * psi); digits = 4)
    push!(bitstrings, (p, bitstring))
end
sort(bitstrings; rev=true)

512-element Vector{Tuple{Any, Any}}:
 (0.992, "101010101")
 (0.0029, "010011001")
 (0.0017, "001011001")
 (0.0007, "010011010")
 (0.0006, "100010101")
 (0.0005, "101010100")
 (0.0002, "101010001")
 (0.0002, "101000101")
 (0.0001, "101100100")
 (0.0001, "100000100")
 (0.0001, "011011001")
 (0.0001, "010111010")
 (0.0001, "001010101")
 ⋮
 (0.0, "000001011")
 (0.0, "000001010")
 (0.0, "000001001")
 (0.0, "000001000")
 (0.0, "000000111")
 (0.0, "000000110")
 (0.0, "000000101")
 (0.0, "000000100")
 (0.0, "000000011")
 (0.0, "000000010")
 (0.0, "000000001")
 (0.0, "000000000")

In [48]:
unzip(a) = map(x->getfield.(a, x), fieldnames(eltype(a)))

total_probability = sum(unzip(bitstrings)[1])

0.9999999989999999

# Applications: Maximum Independent Set (MIS)problems


Consider an undirected graph composed of a set of vertices connected by unweighted edges. An independent set of this graph is a subset of vertices where any two elements of this subset are not connected by an edge. The Maximum Independent Set (MIS) corresponds to the largest of such subsets, and it is in general an NP-complete problem to determine the MIS of a graph.

For example, assume an ensemble of identical radio transmitters over French cities that each have the same radius of transmission. It was quickly realized that two transmitters with close or equal frequencies could interfere with one another, hence the necessity to assign non-interfering frequencies to overlapping transmiting towers. Because of the limited amount of bandwidth space, some towers have to be assigned the same or close frequencies. The MIS of a graph of towers indicate the maximum number of towers that can have close or equal given frequency (red points).

![MIS radio transmitters](./MIS_radio_transmitters.png)


We can attribute a status 
$z$ to each node, where $z_i = 1$ if node $i$ is attributed to the independent set, and 
$z_i = 0$ otherwise. The Maximum Independent Set corresponds to the minima of the following cost function:

$$C(z) = -\sum z_i + U\sum\limits_{\langle ij \rangle} z_i z_j $$


#### Small example
A system with 3 interconnected nodes has following configuration $\{000, 001, 010, 011, 100, 101, 110, 111\}$

Two-body term $C_2(z) = U\sum\limits_{\langle ij \rangle} z_i z_j$ "counts" number of connected nodes in a set

$$C_2(000) = C_2(001) = C_2(010) = C_2(100) = 0,$$

$$C_2(011) = C_2(101) = C_2(110) = U,$$

$$C_2(111) = 3U.$$

One-body term $C_1(z) = -\sum z_i$ counts number of nodes in a set

$$C_1(000) = 0,$$ 

$$C_1(001) = C_1(010) = C_1(100) = -1,$$

$$C_1(011) = C_1(101) = C_1(110) = -2,$$

$$C_1(111) = -3.$$

and, therefore the full cost function $C(z) = C_1(z) + C_2(z)$ minimazes(nulls) number of pairs and maximizes number of nodes

$$C(001) = C(010) = C(100) = -1,$$

$$C(000) = 0,$$ 

$$C(011) = C(101) = C(110) = U-2,$$

$$C(111) = 3(U - 1).$$

For $U >> 1$ minimum for $C(z)$ is acchieved by $\{100, 010, 001\}$


## Quantum version


Interestingly, the operator $\hat{C}(z)$ associated with the cost function of the previous equation can be natively realized on a neutral atom platform with some constraints on the graph edges. We map a ground state and a Rydberg state of each atom to a two-level system, where 
$$|1⟩ =|r⟩$$ is a Rydberg state and 
$$|0⟩=|g⟩$$ is a ground state. 
An atom in a Rydberg state has an excited electron with a very high principal quantum number and therefore exhibits a huge electric dipole moment. As such, when two atoms are excited to Rydberg states, they exhibit a strong van der Waals interaction $U>>1$. Placing 
$N$ atoms at positions 
$r_j$
in a 2D plane, and coupling the ground state 
$|0⟩$
to the Rydberg state 
$|1⟩$
with a laser system enables the realization of the Hamiltonian :
$$\hat{C} = H = \sum\limits_{i=1}^{N} \frac{\Omega}{2} \sigma^x_i - \sum\limits_{i=1}^{N} \frac{\delta}{2}\sigma^z_i + \sum\limits_{i<j} \frac{C_6}{|r_i-r_j|^6} n_i n_j$$